# **<center>Analysis of online shoppers data**

DataLink:-</br>
https://drive.google.com/file/d/1PYVO_O8fWzv3xO_lAe1tUx9zFydMciw8/view?usp=share_link</br>
Data Description: -</br>
https://drive.google.com/file/d/1pj8PF9YmkYvXqDmJ3lUorbHhys8SB0PR/view?usp=share_link


**Questions:**

**1.** Gather and prepare data: Gather and prepare the data you will use to train the model.
This may involve collecting data from various sources, do EDA, cleaning and formatting
the data, and splitting it into training, validation, and testing sets.


**Data Collection**

In [ ]:
# Download file from link)
# in given link, id = 1PYVO_O8fWzv3xO_lAe1tUx9zFydMciw8

!gdown https://drive.google.com/uc?id=1PYVO_O8fWzv3xO_lAe1tUx9zFydMciw8

Downloading...
From: https://drive.google.com/uc?id=1PYVO_O8fWzv3xO_lAe1tUx9zFydMciw8
To: /content/online_shoppers_intention.csv
100% 1.07M/1.07M [00:00<00:00, 67.1MB/s]


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [ ]:
df = pd.read_csv("/content/online_shoppers_intention.csv")
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

**EDA**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

In [ ]:
df.nunique()

Administrative               27
Administrative_Duration    3335
Informational                17
Informational_Duration     1258
ProductRelated              311
ProductRelated_Duration    9551
BounceRates                1872
ExitRates                  4777
PageValues                 2704
SpecialDay                    6
Month                        10
OperatingSystems              8
Browser                      13
Region                        9
TrafficType                  20
VisitorType                   3
Weekend                       2
Revenue                       2
dtype: int64

In [ ]:
df.isna().sum()

# There is no null values

Administrative             0
Administrative_Duration    0
Informational              0
Informational_Duration     0
ProductRelated             0
ProductRelated_Duration    0
BounceRates                0
ExitRates                  0
PageValues                 0
SpecialDay                 0
Month                      0
OperatingSystems           0
Browser                    0
Region                     0
TrafficType                0
VisitorType                0
Weekend                    0
Revenue                    0
dtype: int64

In [ ]:
df["Revenue"].value_counts()

# Data is biased we have to make it balance by using smote technique

False    10422
True      1908
Name: Revenue, dtype: int64

In [ ]:
df.select_dtypes(exclude=(int,float)).columns

Index(['Month', 'VisitorType', 'Weekend', 'Revenue'], dtype='object')

**Data Cleaning**

In [ ]:
# Encoding to object data type columns
df["Month"].replace({"Feb":2,"Mar":3,"May":5, 'Oct':10, 'June':6, 'Jul':7, 'Aug':8, 'Nov':11, 'Sep':9, 'Dec':12}, inplace=True)

df["VisitorType"].replace({'Returning_Visitor':0, 'New_Visitor':1, 'Other':2}, inplace=True)

In [ ]:
df["Revenue"].value_counts()

# Data is biased

False    10422
True      1908
Name: Revenue, dtype: int64

In [ ]:
x = df.drop(["Revenue"], axis=1)
y = df["Revenue"]

In [ ]:
# As Data is biased, so we have to make it balance by using smote technique

from pandas.core.common import random_state
from imblearn.over_sampling import SMOTE
smt = SMOTE()
x_sample, y_sample = smt.fit_resample(x,y)
y_sample.value_counts()

# Now data is Balance

False    10422
True     10422
Name: Revenue, dtype: int64

**Train Test Split**

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x_sample,y_sample,test_size=0.2, random_state=42)

In [ ]:
x_train.shape

(16675, 17)

In [ ]:
x_test.shape

(4169, 17)

**2. Choose a model: Use Decision Tree Algorithm.**

**Create Model**

In [ ]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier()

**3. Train the model: You'll need to train it using the data you collected and prepared in step 2. This involves feeding the data into the model and adjusting the model's parameters until it produces accurate results.**

In [ ]:
model.fit(x_train,y_train)
y_pred_test = model.predict(x_test)
y_pred_train = model.predict(x_train)

**4. Evaluate the model: After training the model, you'll need to evaluate its performance using the validation and testing sets. This will help you determine how well the model is able to generalize to new data and whether it is overfitting or underfitting.**

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
print(f"Testing Accuracy is: {accuracy_score(y_pred_test,y_test)}")
print(f"Training Accuracy is: {accuracy_score(y_pred_train,y_train)}")

Testing Accuracy is: 0.8970976253298153
Training Accuracy is: 1.0


#####Training Accuracy of model is 1 and testing accuracy is 0.89, Means our model is get overfitted

In [ ]:
# Training Performance

cnf_mat = confusion_matrix(y_pred_train,y_train)
accuracy = accuracy_score(y_pred_train,y_train)
report = classification_report(y_pred_train,y_train)

print(f"confusion_matrix: \n{cnf_mat}")
print(f"Training Accuracy is: {accuracy}")
print(f"Classification Report is: \n{report}")

confusion_matrix: 
[[8360    0]
 [   0 8315]]
Training Accuracy is: 1.0
Classification Report is: 
              precision    recall  f1-score   support

       False       1.00      1.00      1.00      8360
        True       1.00      1.00      1.00      8315

    accuracy                           1.00     16675
   macro avg       1.00      1.00      1.00     16675
weighted avg       1.00      1.00      1.00     16675



In [ ]:
# Testing Performance

cnf_mat = confusion_matrix(y_pred_test,y_test)
accuracy = accuracy_score(y_pred_test,y_test)
report = classification_report(y_pred_test,y_test)

print(f"confusion_matrix: \n{cnf_mat}")
print(f"Testing Accuracy is: {accuracy}")
print(f"Classification Report is: \n{report}")

confusion_matrix: 
[[1825  192]
 [ 237 1915]]
Testing Accuracy is: 0.8970976253298153
Classification Report is: 
              precision    recall  f1-score   support

       False       0.89      0.90      0.89      2017
        True       0.91      0.89      0.90      2152

    accuracy                           0.90      4169
   macro avg       0.90      0.90      0.90      4169
weighted avg       0.90      0.90      0.90      4169



In [ ]:
# for training data Precision, recall and accuracy is 1
# for testing data accuracy is 0.89
# Model is overfitted

**5. Fine-tune the model: Based on the results of the evaluation, you may need to fine-tune the model by adjusting the model's parameters, selecting different features, or using a different model architecture.**

In [ ]:
# Parameter to be tune
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [2, 4, 6, 8, 10],
    'min_samples_leaf': [1, 2, 4, 8],
    'min_samples_split': [2, 4, 8, 16]
}

In [ ]:
# hyper parameter tuning by grid GridSearchCV

from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(model,param_grid, cv=5,scoring='accuracy', n_jobs=-1)
grid_search.fit(x_train,y_train)
print("Best hyperparameters:", grid_search.best_params_)
print("Best accuracy score:", grid_search.best_score_)

Best hyperparameters: {'criterion': 'gini', 'max_depth': 8, 'min_samples_leaf': 2, 'min_samples_split': 2}
Best accuracy score: 0.8992503748125937


In [ ]:
# Fit the model
new_model = grid_search.best_estimator_
new_model.fit(x_train,y_train)
y_pred_test = new_model.predict(x_test)
y_pred_train = new_model.predict(x_train)

In [ ]:
# Training Performance after Hyperparameter tuning

cnf_mat = confusion_matrix(y_pred_train,y_train)
accuracy = accuracy_score(y_pred_train,y_train)
report = classification_report(y_pred_train,y_train)

print(f"confusion_matrix: \n{cnf_mat}")
print(f"Training Accuracy is: {accuracy}")
print(f"Classification Report is: \n{report}")

confusion_matrix: 
[[7681  708]
 [ 679 7607]]
Training Accuracy is: 0.9168215892053972
Classification Report is: 
              precision    recall  f1-score   support

       False       0.92      0.92      0.92      8389
        True       0.91      0.92      0.92      8286

    accuracy                           0.92     16675
   macro avg       0.92      0.92      0.92     16675
weighted avg       0.92      0.92      0.92     16675



In [ ]:
# Testing Performance after Hyperparameter tuning

cnf_mat = confusion_matrix(y_pred_test,y_test)
accuracy = accuracy_score(y_pred_test,y_test)
report = classification_report(y_pred_test,y_test)

print(f"confusion_matrix: \n{cnf_mat}")
print(f"Testing Accuracy is: {accuracy}")
print(f"Classification Report is: \n{report}")

confusion_matrix: 
[[1852  191]
 [ 210 1916]]
Testing Accuracy is: 0.9038138642360278
Classification Report is: 
              precision    recall  f1-score   support

       False       0.90      0.91      0.90      2043
        True       0.91      0.90      0.91      2126

    accuracy                           0.90      4169
   macro avg       0.90      0.90      0.90      4169
weighted avg       0.90      0.90      0.90      4169



After hyperparameter tuning Training Accuracy is 91.68 with precision and recall is 0.92</br>
And Testing Accuracy is 90.38 with precision is 0.90 and recall is 0.91</br>
Now model is free from Overfitting